Preliminary implementation of default values

nshaffer opened the pull request:
In support of #62
 
- New module "default_m" which exports the generic function name "default"
- New entries in src/CMakeLists.txt and src/Makefile.manual

I've implemented a generic interface for `real(sp)`, `real(dp)`, `real(qp)`, `int16`, `int32`, `int64`, `logical`, and `character`. I figure this is enough for a first pass. Complex types and discriminating between ASCII and UCS characters can come later. Probably no need to do specific logical kinds.

What's missing right now is tests. I don't know how to use CTest, but I would like to learn if anyone can link me a decent tutorial.

jvdp1 commented on src/stdlib_experimental_default.f90:
```diff
@@ -0,0 +1,142 @@
+module default_m
```
```suggestion
module stdlib_experimental_default
```

jvdp1 commented on src/stdlib_experimental_default.f90:
```diff
@@ -0,0 +1,142 @@
+  !! For additional clarity, `default` be called with keyword argument
+  !! for the fallback value, e.g., `default(x, to=1.0)`.
+  !!
+  use iso_fortran_env, only: sp => real32, dp => real64, qp => real128, int16, int32, int64
```
At some points, it should use `stdlib_experimental_kinds` proposed in #63 . But I would say it is fine for now.

jvdp1 commented on src/stdlib_experimental_default.f90:
```diff
@@ -0,0 +1,142 @@
+     module procedure default_character
+     ! TODO: complex kinds
+     ! TODO: differentiate ascii & ucs char kinds
+  end interface default
```
These lines show again the importance to find a solution to #35 soon, to avoid to go one more time across all files when #35 is solved.

jvdp1 commented on src/stdlib_experimental_default.f90:
```diff
@@ -0,0 +1,142 @@
+     module procedure default_sp
+     module procedure default_dp
+     module procedure default_qp
+     module procedure default_int16
```
`int8` should be added too.

nshaffer replied:
Done, cheers.

jvdp1 commented on src/stdlib_experimental_default.f90:
```diff
@@ -0,0 +1,142 @@
+    end if
+  end function default_character  
+  
+end module default_m
```
```suggestion
end module stdlib_experimental_default
```

certik commented:
Thanks @nshaffer. I resolved the conflict with master and committed to your branch. Make sure you pull it before adding more commits.

Also, I recommend people to submit PRs from a different branch than their master, so that they can use their master to track the upstream master (to rebase, merge, etc.).

certik commented:
Regarding the test, try to copy let's say the `test_ascii` test and its CMake setup. It should be straightforward. If you have any questions, let us know.

nshaffer commented:
Thanks all for the comments. Seems like folks in #62 prefer the name "optval" so I converted to that everywhere and renamed the fallback argument "default", so that you can write either `optval(x, default=1.0)` if you want to be explicit. Also wrote tests for all currently implemented types.

certik commented:
Thank you! This is great. I'll review later tonight or tomorrow.

On Fri, Jan 3, 2020, at 5:59 PM, nshaffer wrote:
> Thanks all for the comments. Seems like folks in #62 
> <https://github.com/fortran-lang/stdlib/issues/62> prefer the name 
> "optval" so I converted to that everywhere and renamed the fallback 
> argument "default", so that you can write either `optval(x, 
> default=1.0)` if you want to be explicit. Also wrote tests for all 
> currently implemented types.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/fortran-lang/stdlib/pull/73?email_source=notifications&email_token=AAAFAWAYRYK3SLJO6IHLFUDQ37NQPA5CNFSM4KCPS462YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEICNLBI#issuecomment-570742149>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAFAWDH262SE7MZ3L5F6OLQ37NQPANCNFSM4KCPS46Q>.
>

certik commented:
Thanks @nshaffer for the work!

certik commented:
Here is the first usage of `optval`: https://github.com/fortran-lang/stdlib/pull/71/commits/a8416a10a1b2db974dfa6192fe486c8d16b24e8c

nshaffer commented:
Cheers! 

It occurs to me that `optval` should be `pure elemental` for the intrinsic types. Shall I tack that on here or open a new PR? (I've added a dev branch to my fork now, if that makes any difference.)

certik commented:
Yes, send another PR please against master, from a new branch.
